packages/flutter/lib/src/widgets/sliver_coordinator.dart

```diff
@@ -0,0 +1,270 @@
+    return _get<_SliverLayoutInfo>(id) != null;
+  }
+
+  /// The [SliverConstraints] computed for this sliver in the most recent
+  /// [CustomScrollView] layout.
+  SliverConstraints getSliverConstraints(Object id) {
+    assert(hasLayoutInfo(id));
+    return _get<_SliverLayoutInfo>(id)!.constraints;
+  }
+
+  /// The [SliverGeometry] computed for this sliver in the most recent
+  /// [CustomScrollView] layout.
+  SliverGeometry getSliverGeometry(Object id) {
+    assert(hasLayoutInfo(id));
+    return _get<_SliverLayoutInfo>(id)!.geometry;
+  }
+}
+
+class _SliverCoordinatorScope extends InheritedWidget {
+  const _SliverCoordinatorScope({ required this.data, required super.child });
+
+  final SliverCoordinatorData data;
+
+  @override
+  bool updateShouldNotify(_SliverCoordinatorScope oldWidget) => false;
+}
+
+/// Provides [ScrollNotification]s for its [CustomScrollView] descendant as well as
+/// layout information for each [CoordinatedSliver] contained by the scroll view.
+///
+/// This widget is intended for triggering animations or auto-scrolls in reaction to
+/// scroll notifications and the concomittant layout changes in the scroll view's
+/// [CoordinatedSliver]s. It cannot be used to change sliver layouts. Typical use cases:
+///
+/// - When a drag gesture ends: auto-scrolling a widget that overlaps the top
+/// or bottom of the viewport so that it's completely visible.
+/// - When a sliver scrolls under a pinned header, change the header's
+/// content or elevation with an animated transition.
+/// - Show an indicator while a special sliver is visible or when its
+/// visibility changes.
+///
+/// {@tool dartpad}
+/// This example contains one [CoordinatedSliver] which is
+/// auto-scrolled so that it's aligned with the top of the viewport
+/// whenever a scroll-end gesture leaves it partially visible. The
+/// auto-scroll is triggered by the [SliverCoordinator]'s
+/// callback. The callback has access to the current scroll offset and
+/// the current extent (height) of the [CoordinatedSliver].
+///
+/// ** See code in examples/api/lib/widgets/sliver/sliver_coordinator.0.dart **
+/// {@end-tool}
+class SliverCoordinator extends StatefulWidget {
+  /// Create specialized scroll notification listener for a [CustomScrollView] descendant.
+  ///
+  /// The [callback] runs each time the scroll view is scrolled.
+  const SliverCoordinator({ super.key, required this.callback, required this.child });
+
+  /// The widget below this widget in the tree.
+  ///
+  /// {@macro flutter.widgets.ProxyWidget.child}
+  final Widget child;
+
+
+  /// Called each time the [CustomScrollView] below this widget scrolls.
+  ///
+  /// This callback runs after its descendants, including the scroll
+  /// view, have been laid out.  The callback's
+  /// [SliverCoordinatorData] parameter includes the
+  /// [SliverConstraints] and [SliverGeometry] computed for each
+  /// [CoordinatedSliver].
+  final SliverCoordinatorCallback callback;
```

Comment:
Can we call this something that indicates when it is called? Like maybe onScrollNotification, instead of just callback?

Reply:
Yes, `onScrollNotification` would be much clearer.